src/main/java/com/google/cloud/spanner/pgadapter/utils/Converter.java

```diff
@@ -27,11 +29,62 @@
 import com.google.cloud.spanner.pgadapter.parsers.NumericParser;
 import com.google.cloud.spanner.pgadapter.parsers.StringParser;
 import com.google.cloud.spanner.pgadapter.parsers.TimestampParser;
+import com.google.cloud.spanner.pgadapter.statements.IntermediateStatement;
 import com.google.common.base.Preconditions;
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
 
 /** Utility class for converting between generic PostgreSQL conversions. */
 public class Converter {
-  private Converter() {}
+  private final ByteArrayOutputStream buffer = new ByteArrayOutputStream(256);
+  private final DataOutputStream outputStream = new DataOutputStream(buffer);
+  private final IntermediateStatement statement;
+  private final QueryMode mode;
+  private final OptionsMetadata options;
+  private final ResultSet resultSet;
+
+  public Converter(
+      IntermediateStatement statement,
+      QueryMode mode,
+      OptionsMetadata options,
+      ResultSet resultSet) {
+    this.statement = statement;
+    this.mode = mode;
+    this.options = options;
+    this.resultSet = resultSet;
+  }
+
+  public int getColumnCount() {
+    return resultSet.getColumnCount();
+  }
+
+  public ResultSet getResultSet() {
+    return resultSet;
+  }
+
+  public int convertResultSetRowToDataRowResponse() throws IOException {
+    buffer.reset();
+    outputStream.writeShort(resultSet.getColumnCount());
+    for (int column_index = 0; /* column indices start at 0 */
+        column_index < resultSet.getColumnCount();
+        column_index++) {
+      if (resultSet.isNull(column_index)) {
+        outputStream.writeInt(-1);
+      } else {
+        DataFormat format = DataFormat.getDataFormat(column_index, statement, mode, options);
+        byte[] column = Converter.convertToPG(this.resultSet, column_index, format);
+        outputStream.writeInt(column.length);
+        outputStream.write(column);
+      }
+    }
+    return buffer.size();
+  }
+
+  public void writeBuffer(DataOutputStream outputStream) throws IOException {
+    buffer.writeTo(outputStream);
+  }
```

Comment:
Why are we using `buffer` while we are not setting it anywhere? Do we mean `this.outputStream.writeTo(outputStream);` ?

Reply:
See also explanation above.

`buffer` is created when a `Converter` is created (see https://github.com/GoogleCloudPlatform/pgadapter/blob/cfb7cd2925e8d5d828b3e65216400edb7e3c7889/src/main/java/com/google/cloud/spanner/pgadapter/utils/Converter.java#L40). This buffer is reused for each row that we convert, and automatically allocates more space as is needed to convert a given row.

Reply:
Ah, yes, good question. Yes, what is happening here is the following:
1. We initialize `buffer` on [line 40](https://github.com/GoogleCloudPlatform/pgadapter/blob/cfb7cd2925e8d5d828b3e65216400edb7e3c7889/src/main/java/com/google/cloud/spanner/pgadapter/utils/Converter.java#L40) with an initial buffer capacity of 256 bytes. That will grow dynamically if needed.
2. We initialize `outputStream` on [line 41](https://github.com/GoogleCloudPlatform/pgadapter/blob/cfb7cd2925e8d5d828b3e65216400edb7e3c7889/src/main/java/com/google/cloud/spanner/pgadapter/utils/Converter.java#L41) with `buffer` as its input argument. `outputStream` is of type `DataOutputStream` and requires another output stream where it can write its data. A `DataOutputStream` is just a convenient way of writing data using native Java types (e.g. writing something like an `int`). The `DataOutputStream` will convert this into bytes and write it to the underlying output stream. In this case, that is our previously defined `buffer`.

So the above means that when we write to `outputStream` in the `convertResultSetRowToDataResponse`, it will automatically convert whatever Java type that we write to it into the corresponding stream of bytes. Those bytes are then written to the underlying `buffer`.